coredns/plugin/k8s.go

```diff
@@ -0,0 +1,213 @@
+package kuadrant
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	core "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+
+	"github.com/kuadrant/coredns-kuadrant/dnsop"
+)
+
+const (
+	defaultResyncPeriod   = 0
+	watchNamespacesEnvVar = "WATCH_NAMESPACES"
+	zoneNameLabel         = "kuadrant.io/zone-name"
+)
+
+type zoneInformers struct {
+	informers  []cache.SharedInformer
+	zone       *Zone
+	zoneOrigin string
+}
+
+func (zi *zoneInformers) refreshZone() {
+	log.Infof("updating zone %s", zi.zoneOrigin)
+	newZ := NewZone(zi.zoneOrigin)
+
+	for _, informer := range zi.informers {
+		for _, obj := range informer.GetStore().List() {
+			rec := obj.(*v1alpha1.DNSRecord)
+			for _, ep := range rec.Spec.Endpoints {
+				log.Debugf("adding %s record endpoints %s to zone %s", ep.RecordType, ep.DNSName, zi.zoneOrigin)
+				err := newZ.InsertEndpoint(ep)
+				if err != nil {
+					log.Error(err)
+				}
+			}
+		}
+	}
+
+	zi.zone.RefreshFrom(newZ)
+}
+
+// KubeController stores the current runtime configuration and cache
+type KubeController struct {
+	client      dnsop.Interface
+	controllers []zoneInformers
+	hasSynced   bool
+	labelFilter string
+}
+
+func newKubeController(ctx context.Context, c *dnsop.DNSRecordClient, zones map[string]*Zone) *KubeController {
+	ctrl := &KubeController{
+		client: c,
+	}
+
+	if existDNSRecordCRDs(ctx, c) {
+		for origin, zone := range zones {
+			labelSelector := labels.SelectorFromSet(map[string]string{
+				zoneNameLabel: stripClosingDot(origin),
+			})
+
+			var namespaces []string
+			if w := os.Getenv(watchNamespacesEnvVar); w != "" {
+				namespaces = strings.Split(w, ",")
+			} else {
+				namespaces = []string{core.NamespaceAll}
+			}
+
+			log.Infof("creating zone informer for %s with label selector %s and namespaces %s", origin, labelSelector.String(), namespaces)
+
+			zi := zoneInformers{
+				informers:  make([]cache.SharedInformer, 0, len(namespaces)),
+				zone:       zone,
+				zoneOrigin: origin,
+			}
+
+			for _, ns := range namespaces {
+				informer := cache.NewSharedInformer(
+					&cache.ListWatch{
+						ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+							opts.LabelSelector = labelSelector.String()
+							return c.DNSRecords(ns).List(ctx, opts)
+						},
+						WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+							opts.LabelSelector = labelSelector.String()
+							return c.DNSRecords(ns).Watch(ctx, opts)
+						},
+					},
+					&v1alpha1.DNSRecord{},
+					defaultResyncPeriod,
```

Comment:
does it matter that we are passing 0 as default? Not sure what that means

Reply:
AFAIK it just means that it won't resync when it's set to 0, i believe this is the default so could just be removed and it would do the same thing. https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L323